Update commit ID in rpm specification files when updating commit ID i…

alanoe opened the pull request:
…n yaml files in upgrade_versions.py

Commit:
Update commit ID in rpm specification files when updating commit ID in yaml files

This is done in upgrade_versions.py script.

olavphibm commented on tools/upgrade_versions.py:
```diff
@@ -61,6 +61,16 @@ def _sed_yaml_descriptor(yamlfile, old_commit, new_commit):
+    with file(specfile, "w") as f:
+        for line in lines:
+            line = line.replace(old_commit, new_commit)
+            f.write(line)
```
This function is identical to the one above, you could just change its variables to have more generic names.

alanoe replied:
Done

olavphibm commented on tools/upgrade_versions.py:
```diff
@@ -132,6 +142,8 @@ def _bump_release(self, pkg, change_log_header=None, user_name=None,
             _sed_yaml_descriptor(
                 self.pkg.package_file, old_commit_id, new_commit_id)
+            _sed_rpm_spec(
+                self.pkg.package_file, old_commit_id, new_commit_id)
```
You're passing the yaml file to both functions.

alanoe replied:
Done

mopsfelder commented:
Perhaps we can have a more generic function, say `lib.utils.inplace_replace()`, that can be used in `_sed_yaml_descriptor()` and `_sed_rpm_spec()`.  For example:

```python
import fileinput
def inplace_replace(files, pattern, replace):
  for line in fileinput.input(files=files, inplace=True):
    print(line.replace(pattern, replace))

def _sed_yaml_descriptor(yamlfile, old_commit, new_commit):
  inplace_replace(file=yamlfile, old_commit, new_commit)

def _sed_rpm_spec(specfile, old_commit, new_commit):
  inplace_replace(file=specfile, old_commit, new_commit)
```

ghost commented:
I like @mopsfelder's `inplace_replace` function, I just don't see a reason to keep the identical wrapper functions.

alanoe commented:
OH> Please try to keep short commit messages shorter than 72 characters, so they fit in the PR and commit titles.

We already have layout rules for Python code and commit message (assuming we have inherited the ones we used in PowerKVM), do we need to add an extra one? I personally always prefer a longer and clearer description than a non-intuitive one. It makes easier to anyone inside or outside the team to understand what the patch does without reading its content.

OH> Perhaps we can have a more generic function, say lib.utils.inplace_replace(), that can be used in _sed_yaml_descriptor() and _sed_rpm_spec()

Done.  I just used a more intuitive name. I agree with Olav that the wrapper functions are useless.

Commit:
Improve names of methods which update release number and change log in rpm spec file

- There is an update_version() method in SpecFile class in rpm_package.py which
  increments the version number, thus the method which increments the release number
  should use 'update' instead of 'bump' for consistency
- Old bump_release() method also updates rpm spec change log, this should be
  obvious in the the method name

ghost commented:
Most sources that I've found (including git documentation: https://git-scm.com/docs/git-commit#_discussion) recommend writing the short commit message with at most 50 characters. I personally find it hard to do this with many commits, so I'm just suggesting we keep a maximum of 72 to avoid breaking readability in GitHub like this:
```
Update commit ID in rpm specification files when updating commit ID i…

…n yaml files
```
This line break also appears in the PR title which is then used as the default commit message when merging the PR.

The longer commit message can be used to explain the commit in detail.

olavphibm commented on tools/upgrade_versions.py:
```diff
@@ -104,9 +104,9 @@ def update(self, user_name, user_email):
 
-    def _bump_release(self, pkg, change_log_header=None, user_name=None,
+    def _update_release_and_change_log(self, pkg, change_log_header=None, user_name=None,
                       user_email=None):
```
Indentation became wrong.

farosas commented on lib/rpm_package.py:
```diff
@@ -103,7 +103,7 @@ def update_version(self, new_version):
         self.write_content()
 
-    def bump_release(self, change_log_lines, user_name, user_email):
+    def update_release_and_change_log(self, change_log_lines, user_name, user_email):
```
Historical information: 'bump' is not the same as 'update' because bump is usually used to mean 'increment'. I used 'update' because the version might be totally wrong since we don't have control over what the package owners do.

Regardless, I don't see how bumping the release would not cause the change log to be updated or how updating the change log would not cause the release to be bumped.

alanoe replied:
FR> Regardless, I don't see how bumping the release would not cause the change log to be updated or how updating the change log would not cause the release to be bumped.

If the rpm package does not keep a change log, you could simply increment the release number. Also, it is not obvious to a beginner that a rpm spec even has a change log. Several other file descriptors (eg XML/JSON/YAML files, such as our package metadata yaml file) do not. I always recommend not making assumptions and keeping things explicit.

olavphibm commented on tools/upgrade_versions.py:
```diff
@@ -128,7 +128,7 @@ def _bump_release(self, pkg, change_log_header=None, user_name=None,
             assert user_name is not None
             assert user_email is not None
-            pkg.spec_file.bump_release(change_log_lines, user_name, user_email)
+            pkg.spec_file.update_release_and_change_log(change_log_lines, user_name, user_email)
```
Line is too long.

olavphibm commented on lib/utils.py:
```diff
@@ -126,3 +126,14 @@ def run_command(cmd, **kwargs):
+    with file(file_path, "w") as f:
+        for line in lines:
+            line = line.replace(search, replacement)
+            f.write(line)
```
What about @mopsfelder's suggestion of using fileinput here? It simplifies this function.

alanoe replied:
Using fileinput reduces code size, but I personally dislike the print() in it. I had to open Python std lib doc to understand that "the file is moved to a backup file and standard output is directed to the input file" in it. To me, the code above is simpler to understand. But if you guys prefer using fileinput, I can use it.

olavphibm replied:
I see your point. I'm ok with leaving it like this then.

alanoe commented:
OH> I'm just suggesting we keep a maximum of 72 to avoid breaking readability in GitHub like this

I have understood the motivation, but I think this issue is *very* minor because you'll need to open the PR to look at the patch anyway and then you'll see the full commit msg. Besides that, I think describing the patch clearly is much more important than having a concise commit msg. Anyway, for most cases, using the longer commit msg to describe the patch better works. I just wished we spent more time discussing algorithms, design and performance than code/doc layout in code reviews.
